File: python/examples/spdp_announce.py
#!/usr/bin/env python3
"""Standalone SPDP announcer — sends participant announcements via multicast.

Usage:
    python examples/spdp_announce.py [--domain DOMAIN_ID]

Announce a VibeDDS participant on the DDS domain. Other DDS implementations
(e.g., rtiddsspy) should see this participant appear.
"""

import sys
import os
import time
import argparse
import logging

# Add parent directory to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from vibedds.participant import DomainParticipant


def main():
    parser = argparse.ArgumentParser(description="SPDP Announcer")
    parser.add_argument("--domain", type=int, default=0, help="DDS domain ID")
    parser.add_argument("--interval", type=float, default=5.0,
                        help="Announce interval in seconds")
    parser.add_argument("-v", "--verbose", action="store_true")
    args = parser.parse_args()

    logging.basicConfig(
        level=logging.DEBUG if args.verbose else logging.INFO,
        format="%(asctime)s %(name)s %(levelname)s: %(message)s",
    )

    participant = DomainParticipant(domain_id=args.domain)
    participant._spdp_announce_interval = args.interval
    participant.start()

    print(f"SPDP Announcer started on domain {args.domain}")
    print(f"  GUID Prefix: {participant.guid_prefix}")
    print(f"  Local IP: {participant.transport.local_ip}")
    print(f"  Metatraffic port: {participant.transport.metatraffic_unicast_port}")
    print(f"  User data port: {participant.transport.user_unicast_port}")
    print(f"  Announce interval: {args.interval}s")
    print("Press Ctrl+C to stop.\n")

    # Send first announcement immediately
    participant.announce_spdp()

    try:
        while True:
            participant.spin_once(timeout=0.5)
    except KeyboardInterrupt:
        print("\nStopping...")
    finally:
        participant.stop()


if __name__ == "__main__":
    main()

File: rust/vibedds/examples/spdp_announce.rs
/// SPDP announce example - sends periodic participant announcements.
///
/// Run with: cargo run --example spdp_announce
///
/// Use Python spdp_listen.py to verify interop:
///   python examples/spdp_listen.py

use std::time::Duration;
use vibedds::participant::DomainParticipant;

fn main() {
    env_logger::Builder::from_env(env_logger::Env::default().default_filter_or("info")).init();

    let mut participant = DomainParticipant::new(0, 0);

    participant.on_participant_discovered(|p| {
        log::info!(
            "Discovered participant: {:?}, vendor={:?}",
            p.guid_prefix,
            p.vendor_id
        );
    });

    participant.on_participant_lost(|p| {
        log::info!("Lost participant: {:?}", p.guid_prefix);
    });

    log::info!("Starting SPDP announcer on domain 0...");
    log::info!("GUID prefix: {:?}", participant.guid_prefix());

    if let Err(e) = participant.start() {
        log::error!("Failed to start participant: {}", e);
        return;
    }

    // Set faster announce interval for testing
    participant.set_spdp_announce_interval(Duration::from_secs(5));

    log::info!("Running for 60 seconds. Use Python spdp_listen.py to see announcements.");

    if let Err(e) = participant.spin(Some(Duration::from_secs(60))) {
        log::error!("Error in spin loop: {}", e);
    }

    participant.stop();
    log::info!("Done.");
}
